src/groups/mqb/mqbnet/mqbnet_tcpsessionfactory.cpp

```diff
@@ -1414,5 +1422,73 @@ bool TCPSessionFactory::isEndpointLoopback(const bslstl::StringRef& uri) const
            mwcio::ChannelUtil::isLocalHost(endpoint.host());
 }
 
+// ------------------------------------
+// class TCPSessionFactory::PortManager
+// ------------------------------------
+
+TCPSessionFactory::PortManager::PortManager(bslma::Allocator* allocator)
+: d_portMap(allocator)
+, d_regex(allocator)
+, d_allocator_p(allocator)
+{
+    const char                  pattern[] = ":(\\d{1,5})$";
+    bsl::string                 error(allocator);
+    size_t                      errorOffset;
+    BSLA_MAYBE_UNUSED const int rc = d_regex.prepare(
+        &error,
+        &errorOffset,
+        pattern,
+        bdlpcre::RegEx::k_FLAG_JIT);
+    BSLS_ASSERT_SAFE(rc == 0);
+    BSLS_ASSERT_SAFE(d_regex.isPrepared());
+}
+
+bslma::ManagedPtr<mwcst::StatContext>
+TCPSessionFactory::PortManager::addChannelContext(mwcst::StatContext* parent,
+                                                  const bsl::string&  endpoint,
+                                                  const bsl::string&  port)
+{
+    bdlma::LocalSequentialAllocator<2048> localAllocator(d_allocator_p);
+    mwcst::StatContextConfiguration statConfig(endpoint, &localAllocator);
```

Comment:
Unfortunately `statConfig` is not the same as it's initialized with different `enpoint` every time. This endpoint is also the name of the StatContext being created below. Hence, it can't be cached. I see two possible solutions in order to move preparation steps out from the lock:

1. To pass a pointer to the `TCPSessionFactory`'s mutex as a parameter to this method (or may be to the `PortManager`'s constructor, saving it as a class member) and lock it a bit later.
2. To add another mutex to this class and use it instead of the `TCPSessionFactory`'s mutex. In this case we can avoid locking the external mutex in both `addChannelContext()`  and `deleteChannelContext` methods

Reply:
Let's skip it for now